docs/extended-collection-types.md

```diff
@@ -0,0 +1,169 @@
+
+`dict.get_value(key)` will return the value of the specified `key` or `null` if the key is not present in the dictionary.
+
+### HashSet
+Essentially identical to an OD list, except it cannot contain duplicates. Inserting an element that is already present in the hashset is not an error and instead does nothing.
+
+Direct insertion (e.g. `hashset += "foo"`) is supported, but `hashset.add(element)` will return `TRUE` or `FALSE` depending on if the element already existed in the hashset.
+
+Example usage:
+```
+var/hashset/foo = hashset("a","b","c")
+
+foo += "c" // Valid but nothing happens since `foo` already contains "c"
+
+foo["c"] = 5 // Compile-time error, hashsets cannot be associative
+
+world.log << foo.add("c") // Prints FALSE and inserts nothing since `foo` already contains "c"
+
+world.log << foo.add("d") // Prints TRUE and inserts "d" into `foo`
+world.log << foo.Join() // Prints "abcd"
+```
+
+### ODList 
+Identical to BYOND `list()` but with no support for associative lists. This facilitates various internal optimizations over using `list()`.
+
+Unlike BYOND lists, ODLists also implement the "Constant Collections" and "Collection Capacity API" defined in subsequent sections.
+
+Example usage:
+```
+var/odlist/foo = odlist("a","b","c")
+
+foo += "d" // Valid modification, just like a BYOND list
+
+foo["c"] = 5 // Compile-time error, ODLists cannot be associative
+
+world.log << foo.Join() // Prints "abcd"
+```
+
+### Queue
+Similar to lists, except elements are first-in first-out. Most of the list API will remain, except for inserting and removing elements. These will use `queue.enqueue(element)` and `queue.dequeue()`.
+
+Example usage:
+```
+var/queue/foo = queue("a","b","c")
+
+foo += "d" // Compile-time error, must use foo.enqueue("d")
+
+world.log << foo[2] // Prints "b", indexing is supported
+
+world.log << foo.dequeue() // Removes "a" from `foo` and prints "a"
+```
+
+A `queue.to_array()` helper will also be provided, and returns the queue's elements as a shallow-copied `array()`.
+
+### Stack 
+Similar to queues, except elements are last-in first-out. Most of the list API will remain, except for inserting and removing elements. These will use `stack.push(element)` and `stack.pop()`.
+
+Example usage:
+```
+var/stack/foo = stack("a","b","c")
+
+foo += "d" // Compile-time error, must use foo.push("d")
+
+world.log << foo[2] // Prints "b", indexing is supported
+
+foo.push("d")
+world.log << foo.pop() // Removes "d" from `foo` and prints "d"
+```
+
+A `stack.to_array()` helper will also be provided, and returns the stack's elements as a shallow-copied `array()`.
+
+## Constant Collections
+
+All OpenDream collections can now be declared as constant, which prevents any future modifications to the collection contents and will enable future internal optimizations.
+
+Example usage:
+```
+var/const/array/foo = array("a","b","c")
+foo[1] = "z" // Compile-time error, cannot modify constant collection
```

Comment:
I think this needs some explanation of how it is going to work with relation to casting.
My assumption is:
- Any non-constant array can be cast into a constant array (or provided as a parameter to a function that takes constant arrays)
- Any constant array cannot be cast into a non-constant array

If the above was the case, then how would the following work?

```
var/array/test = array("a", "b")
var/const/array/constant_test = test
test[1] = "c"
world.log << constant_test[1]
```
Or would constant have to be compile time constant, since if that is the case then I don't really see much use for it?